### ISSUE TITLE
Feature request: Allow registering an event (callback) on test assert failures and exceptions

### ISSUE THREAD

**Issue by cwood-panopto** [nunit/nunit#2990]
Multiple users have asked for some way to trigger some code on test assert failure and exception. See some discussion here https://github.com/nunit/docs/issues/193, here https://stackoverflow.com/questions/9157991/appdomain-unhandledexception-handler-doesnt-fire-in-unit-test, here https://stackoverflow.com/questions/33320912/take-screenshot-on-test-failure-exceptions, and I have seen others but this was my quick google search.

This would allow a number of scenarios to be unlocked. From taking screenshots on failure (there is a large community of Selenium users for this scenario), capturing additional logging or test state, network capture, browser dom or console logs, data from another machine in multi-machine scenarios, etc.

Currently there aren't great ways to do this. TearDown has been suggested but is not always acceptable because it does not always run at the point of failure. Other TearDowns may have occurred. One may also want to capture failure info during TearDown. An assert wrapper has been suggested; this would only allow capturing the assert failures and not exceptions. The wrapper seems more than should be necessary to implement something simple like this.

The most obvious solution seems to be to allow users to register an event or callback that would be ran when a failure or exception occurred. There could even be separate events for assert failures and exceptions.

**Comment by CharliePoole** [comment]
You don't discuss our recommended approach, which is to use an engine extension. Is there some objection to that? I've noticed all the comments you cite and lots more but they all want to take an approach whereby the code to handle test exceptions and errors is actually part of the test. There's a bit of a logical fallacy there. 😄 

A callback or other hook would mean that your code could do __anything__. That seems risky. I'd rather see features that do specific things for a user, such as saving a screen shot. The test would tell NUnit to take the particular action and NUnit would do it in such a way as to protect other tests. If it could't be done, then the call could fail. In general, it seems safer to implement features where the user calls NUnit rather than having NUnit call the user.

**Comment by cwood-panopto** [comment]
I am trying to describe the possible scenarios I have seen people asking for and meet your requirement of opening a new issue as requested even though I think the wish of the original issue poster was to do something on test exceptions (misunderstanding that the original 2.5 framework implementation of exceptions was actually for NUnit "framework" exceptions not exceptions in test code). They even said "I'm just not happy with the default exception output for my integration tests and i want to dump some more information so it's easier to see what the problem is. like in an application AppDomain.UnhandledException += ....". I feel I have brought up the issue well enough at this point as have other users.

I don't really care the exact implementation of this feature. I am just reiterating the details of the issue in this new venue and describing what @jnm2 described in that bug (he mentioned "callback" specifically) and what I feel most people are looking for as I read their discussions online and they have pointed out that other frameworks provide similar solutions but for some reason NUnit does not.

When you are talking about using an engine extension, are you talking about https://github.com/nunit/docs/wiki/Writing-Engine-Extensions? These don't seem well documented. What do you suggest I queue on to do something on test failure or test exception? The Event-Listeners don't seem to include on test failure or exception; as described by the documentation it is only at the end of a run, suite, or test case (https://github.com/nunit/docs/wiki/Event-Listeners). I also recall trying to write some type of extension for this or other purpose and found that they were not compatible or did not play well with the VS Test Adapter or deployment and proliferation of said extension was not simple or consistent between build and dev machines (I don't recall as that was over 6 months ago when I initially started working with NUnit).

If you are suggesting NUnit should take the screenshot itself. That seems highly dependent upon the framework that is driving the browser or app etc (for Selenium, it may not even be the same machine running the test code, for Appium it may be a mobile device attached to a host or virtual device), so I don't see how you could provide that very specific functionality in your framework. Additionally, myself and I believe others want to do things other than just grabbing a screenshot and that may be very individually product specific (so again I don't see how you can provide such a specific thing without calling back to someone's code not written by the framework).

As for the Engine Extension option, if you maybe provide just a little more detail maybe I can continue down that route again and see what the original issue I ran into was. In the end this will only help me and I feel a more holistic solution provided directly by NUnit will solve this problem for a lot more users.

**Comment by jnm2** [comment]
@CharliePoole With an engine extension, how would an NUnit user integrate the various parts?

The benefit of extending the framework instead of the engine is that it doesn't require the end user to control the invocation of the engine. This means you can attach screenshots when running from an IDE.

I was imagining something as simple as this:

```c#
[AttachScreenshots("screenshots")]
public static class SomeTests
{
    [Test]
    public static void SomeTest()
    {
        using (var userInterface = CreateUserInterface())
        using (ScreenshotUtils.SetCurrentWindow(userInterface))
        {
            Assert.That(userInterface.Property, Is.True);
        }
    }
}

// User code
public sealed class AttachScreenshotsAttribute : NewFeature_ApplyToTestCasesAttribute
{
    private readonly string screenshotDirectory;
    private readonly bool failingAssertionsOnly;

    public AttachScreenshotsAttribute(string screenshotDirectory, bool failingAssertionsOnly = true)
    {
        this.screenshotDirectory = screenshotDirectory;
        this.failingAssertionsOnly = failingAssertionsOnly;
    }

    protected override void OnAssertion(AssertionEventArgs e)
    {
        if (failingAssertionsOnly && e.Assertion.Status != AssertionStatus.Failed)
        {
            return;
        }

        ScreenshotUtils.GetScreenshot().Save(
            Path.Combine(screenshotDirectory, e.TestCase.FullName + ".png"),
            ImageFormat.Png);
    }
}

public static class ScreenshotUtils
{
    private static HandleRef currentHWnd;

    public static Bitmap GetScreenshot()
    {
        if (currentHWnd.Handle == IntPtr.Zero) throw new InvalidOperationException("No screenshot test running.");

        // Use currentHWnd to create a bitmap
    }

    public static IDisposable SetCurrentWindow(IWin32Window window)
    {
        return SetCurrentWindow(window, window.Handle);
    }

    public static IDisposable SetCurrentWindow(object wrapper, IntPtr handle)
    {
        if (handle == IntPtr.Zero) throw new ArgumentException("Not a valid handle.", nameof(handle));
        if (currentHWnd.Handle != IntPtr.Zero) throw new InvalidOperationException("Overlapping screenshot tests detected.");

        currentHWnd = new HandleRef(wrapper, handle);
        return On.Dispose(() => currentHWnd = default);
    }
}
```

**Comment by CharliePoole** [comment]
@cwood-panopto I'm only reacting to your post and don't know what others have told you in other issues. Best thing is to describe your need and not try to gather together info from other users' requests, which may sometimes resemble yours superficially but have subtle differences.

The recommended way of __handling__ results after tests are completed is to create an engine extension - unless the team can be convinced to make it a part of NUnit itself. A test listener extension is perfect for that purpose because the feature becomes available to anyone who installs the extension. More important, the test result is provided is called only after the test is completely over and the extension cannot affect it in any way. For reporting purposes, that's a necessary feature, but it's not useful for your purposes.

[BTW, before leaving this, I should let you know that the interface for such an extension is very low-level. You get a report of an event in XML and the extension has to figure out that it's a report of test completion and whether the test failed or passed. It an do anything it wants with the result, but the test is already over by the time it is called.]

Some of the other requests fit this model but you are actually looking to __affect__ the test result. You probably want to record a link to the screen dump in the result itself, but even if you did not, the dump is logically a part of the outcome of the test. For that reason, you need to have something __in__ the test code to do the recording. The rest of the discussion is about how we would do that - what it would look like in an API. It could be a call you make, which means you need to know when to make the call. It could be an attribute as @jnm2 proposed. I'll save commenting on alternative implementations for a separate comment.

**Comment by cwood-panopto** [comment]
I would very much appreciate an ApplyToTestCasesAttribute that could be extended by users which could be applied at a fixture and assembly level. There are many use cases I think this would support which I had tried to solve previously but the only available attributes had to be applied to every single test case.

Would something like this also be able to support test code exceptions and exceptions during SetUp/TearDown?

The other possibly hairy thing seems the user's implementation of something like ScreenshotUtil. For the Selenium case this would require passing the driver object around (much like the window handle in this example) and this seems hairy especially for parallelized test cases. I'd want to avoid every writer of a test case in my organization having to add and know to add a using statement in their test. I think I would add this in a base class we have.
Thanks.

**Comment by CharliePoole** [comment]
@jnm2 I'm pretty much with you on the suggested approach.

Details... details...

I'm a bit stricter to you on what should or should not go in an extension. I agree that writing extensions is harder but that's not a way to distinguish between framework and engine extensions! Rather it's a way to distinguish between writing an extension and __not__ writing one. That's the easiest for most users.

@cwood-panopto Implied that he wanted to write an extension in the wording of his request but I think we should go beyond that. IMO the basic screenshot capability ought to be part of NUnit. However, there are at least two ways to take a screenshot - likely more. Your approach works if there is an actual hwnd available, which I __believe__ is only on windows. For a test that uses some web testing platform you may need to go through the driver. That's where an extension is probably needed. Of course, we don't currently have a way for the framework to load extensions dynamically so it would have to be referenced in the test itself.

**Comment by jnm2** [comment]
@CharliePoole It shouldn't matter whether my example is Win32-based—that's up to the end user to write. The point I'm trying to explore is that the user may want to run from Test Explorer and ReSharper and a CI script. The user could do a small amount of setup with a framework extension point. With an engine extension point, the user must begin controlling distant moving parts in multiple places and it may not even be possible to run the tests properly on all runners.

**Comment by CharliePoole** [comment]
@jnm2 That wasn't clear to me. I assumed it was all internal code except for the attribute, which the user would write. I'm a bit dubious whether we should require the user to write the most difficult code of all. The entire notion of an extension is that the user doesn't have to write it, merely install a package.

WRT engine extensions, they are not suitable for this specific purpose. It has nothing to do with how difficult they are to write, they aren't suitable for anybody to use because they run completely outside the framework and in this case "centuries" after the test has finished. They also run in a completely separate process in most cases! It's just that the bundling of this request with other requests that __are__ engine-extension material confused the issue for me initially.

What would be the nature of the base class attribute you are proposing. Just what would it do. This is a kind of big deal because we have, up to now, __not__ provided base class attributes for extension except in a few cases where, frankly, I think it was a mistake. Our design strategy has been that users should implement interfaces in custom attributes, not inherit from ours. I'm not saying we shouldn't change it... just that it's a significant change in overall approach. What, specifically would the hypothetical attribute do that the user could not do? Also, Why don't we just implement the attribute to take automatic screen dumps ourselves and simply give the user an event like ScreenDumpRequest that they can latch onto?

All that said, if you are planning to write the code, I'm happy to wait and it's better not to discuss it in detail until you do.

**Comment by jnm2** [comment]
@CharliePoole 

> I'm a bit dubious whether we should require the user to write the most difficult code of all.

I'm not saying we shouldn't create something like this. For now, I just want to make sure users will be _able_ to write this kind of code without us having thought of it first. In other words: I think users should be able to write things that react to assertions which either capture screenshots very differently (e.g. Chrome in headless mode) or even do non-screenshot-related tasks.

The big appeal of abstract base classes is that we can suddenly add new virtual methods to them, unlike interfaces. We can operate either way, so that was a straw man. I did it in the example because it seemed neater to not have to specify AttributeUsage in the user code. Not a very important factor.

Let's talk in terms of interfaces since that's the direction NUnit has pursued recently. Something like this:

```c#
public interface IAssertionListener
{
    void OnAssertion(AssertionEventArgs e);
}
```

`AssertionEventArgs` gives us room to grow without breaking changes:

```c#
public sealed class AssertionEventArgs : EventArgs
{
    public ITest Test { get; } // Maybe the in-progress test result?
    public AssertionResult Assertion { get; }
}
```

When implemented by an attribute and applied to a method, class or assembly, NUnit would call into it for all assertions for the test it applies to and all descendant tests.

This hierarchical application would be very similar to the ability being asked in https://github.com/nunit/nunit/issues/2220.

**Comment by CharliePoole** [comment]
@jnm2 I'm happy to see you implement it that way - or anyone else who so chooses. 😄 

As you know, I'm a big opponent of "generalization from one case", which is what I see here, so it's not how I would do it. I wouldn't try to define a system architecture to support __all__ cases where we want to be called from an assertion point in order to simply display screen shots. Nobody has asked for us to take screen shots up to now except on failure or on specific request by the user. I would implement a feature, then I'd implement other features. When I had three features that needed the same structural underpinning, I'd create it and refactor. That's what I would do. I assume that if I were implementing it, you'd let me do it that way. I'll let you do it your way. 😄 

If - on the other hand - we are talking about how some third party has to implement it, let's not do that. Rather, let's wait for the third party to show up and give them advice if they want it.

I'm coming from the perspective that I don't want to over-specify work that will be done by other team members or contributors. Hope that's OK.

**Comment by jnm2** [comment]
@CharliePoole A user has asked for the ability to capture a screenshot at the precise moment of an assertion, but without NUnit making any assumptions about whether the screenshot comes from the desktop, a window, Selenium, a remote device, and so on. 

One option is for NUnit to do nothing and the user to create a `ScreenshotAssert.That` and use it at the point of every test assertion, but users are commonly interested in applying things like this in a cross-cutting way and I do see the appeal to not having to edit every assertion of every test. 

If we decide not to do nothing, what can NUnit reasonably do to satisfy this ask that is any *less* general than providing an assertion hook? The ask is specifically that NUnit does not take the screenshot itself.

**Comment by jnm2** [comment]
@CharliePoole A user has asked for the ability to capture a screenshot at the precise moment of an assertion, but without NUnit making any assumptions about whether the screenshot comes from the desktop, a window, Selenium, a remote device, and so on. 

One option is for NUnit to do nothing and the user to create a `ScreenshotAssert.That` and use it at the point of every test assertion, but users are commonly interested in applying things like this in a cross-cutting way and I do see the appeal to not having to edit every assertion of every test. 

If we decide not to do nothing, what can NUnit reasonably do to satisfy this ask that is any *less* general than providing an assertion hook? The ask is specifically that NUnit does not take the screenshot itself.

> If - on the other hand - we are talking about how some third party has to implement it, let's not do that. Rather, let's wait for the third party to show up and give them advice if they want it.

Any cross-cutting mechanism will mean new user-facing framework APIs, and I do have an interest in making sure that any new APIs are worth their salt no matter what the implementation is like.
> I'm coming from the perspective that I don't want to over-specify work that will be done by other team members or contributors. Hope that's OK.

I appreciate this. Thanks for bearing with my questions as I try to understand.

**Comment by CharliePoole** [comment]
"A user has asked for the ability to capture a screenshot at the precise moment of an assertion, but without NUnit making any assumptions about whether the screenshot comes from the desktop, a window, Selenium, a remote device, and so on."

@cwood-panopto Is that what you are asking for? (Serious question.) If so, do you want to be able to specify it one assert at a time or for all asserts in a test or set of tests? I'm particularly asking because there seem to be two distinct challenges here: the "generality" of how to take a sreen shot and the manner in which we make it available hierarchically.

@jnm2 When I say "challenge" I don't mean I don't know how to do it, but that I know quite a few different ways to do it and the challenge is to pick the right one. If __I__ were programming it, I would not like to be told which to pick. Granted, that's unlikely, but I am trying to put myself in the shoes of whoever does show up to do it.

**Comment by cwood-panopto** [comment]
Specifically I am looking for a mechanism that I can apply to a test fixture (and/or assembly, not just a per test case basis), preferably I would be able to add this mechanism to a SeleniumTestBase class we already have so all our Selenium specific NUnit tests (and not other basic unit tests) may benefit automatically for anyone who writes a Selenium test.
I would like for this mechanism to allow me to capture a screenshot (from local or a remote Selenium Grid's driver object, a physical or virtual mobile device, or other; Selenium already provides the [TakeScreeshot](http://seleniumhq.github.io/selenium/docs/api/dotnet/html/M_OpenQA_Selenium_Support_Extensions_WebDriverExtensions_TakeScreenshot.htm) method), a copy of the DOM from a local or remote browser (Selenium already provides collecting the [PageSource](http://seleniumhq.github.io/selenium/docs/api/dotnet/html/P_OpenQA_Selenium_Remote_RemoteWebDriver_PageSource.htm)), a copy of the console logs from the browser, and a network HAR file for the currently loaded page (I've done this with PhantomJS in the past with Selenium and Python frameworks) very near the point of failure. There are other things I want to collect but they can reasonably be grabbed well after the point of failure, but if this mechanism was provided I don't see why I wouldn't just capture them here for the most filtered timely logs to aid in quicker debugging.
Failures in my mind include test assert failures **AND** unhandled test code exceptions resulting in a test failure from each of SetUp, Test, and TearDown methods possibly even OneTimeTearDown (but OneTimeTearDown would only be a bonus, we only have one suite where the driver is still available at that point in the test).
Near point of failure insinuates that additional user test code has not ran to influence the current state of the browser (for instance, TearDown has not started if the failure was in Test). Sometimes this is a more immediate need though: The browser may be actively loading a page still or executing some javascript or dynamically changing and we want to capture that point in time as closely as reasonably possible for debugging purposes.


Today I used to take screenshots all along the way, during each action that Selenium performs. Selenium provides [Events](http://seleniumhq.github.io/selenium/docs/api/dotnet/html/Events_T_OpenQA_Selenium_Support_Events_EventFiringWebDriver.htm) callbacks for performing custom code during any action. However, each screenshot of a remote machine adds about a second per action to a test without capturing all the other debugging I want to capture (dom, har, console log, etc). Worse Selenium has these Wait objects that throw Timeout exceptions so you can't easily assert on them. And they perform an "action", checking for an element every 500ms, so even a short wait results in tests sometimes taking much longer than 7x their original time which is not acceptable. Even efforts I made to not take screenshots during these waits resulted in test length issues. The most common you would want to take a screenshot for is when Selenium finds an element and this happens all the time (too often, magnitudes more than all the other events combined). Even when I had filtered the set of events where Selenium takes a screenshot (selecting only a few events and excluding the [FindElementCompleted](http://seleniumhq.github.io/selenium/docs/api/dotnet/html/E_OpenQA_Selenium_Support_Events_EventFiringWebDriver_FindElementCompleted.htm) event) resulted in fairly old screenshots where another action had occurred that we had not taken a screenshot for. I don't need all these screenshots all along the way and they take too long, so currently we only collect the DOMs/PageSource and nothing else for a selection of events (as that is order of magnitudes faster than the screenshot).
The reasonable solution to my problem seems to be to only take a screenshot at the point of actual failure as that would only impact each failing test by only 1 second or so (and not every single test by hundreds of seconds). Selenium does not know when a test has "failed" so it seems reasonable this mechanism, on failure, would be provided by NUnit in some way. I can't use many other numerous methods I and others have researched like `AppDomain.UnhandledException += ....` as NUnit captures those.

**Comment by CharliePoole** [comment]
@cwood-panopto Thanks for the clarification. Your comment is very dense so I'm going to quote extensively and interline my own comments.

"Specifically I am looking for a mechanism that I can apply to a test fixture (and/or assembly, not just a per test case basis), preferably I would be able to add this mechanism to a SeleniumTestBase class we already have so all our Selenium specific NUnit tests (and not other basic unit tests) may benefit automatically for anyone who writes a Selenium test."

So, specify once and not within each test. Certainly not within each assert.

"I would like for this mechanism to allow me to capture a screenshot (from local or a remote Selenium Grid's driver object, a physical or virtual mobile device, or other; Selenium provides the TakeScreeshot method), a copy of the DOM from a local or remote browser (Selenium provides collecting the PageSource), a copy of the console logs from the browser, and a network HAR file for the currently loaded page (I've done this with PhantomJS in the past with Selenium and Python frameworks) very near the point of failure. There are other things I want to collect but they can reasonably be grabbed well after the point of failure, but if this mechanism was provided I don't see why I wouldn't just capture them here for the most filtered timely logs to aid in quicker debugging."

Key takeaway here for me is that this sort of thing is outside of the scope of NUnit itself. We need to provide a way that either the user or the vendor of a product like Selenium can integrate and get called at the appropriate time. There could be various providers of this facility: a general desktop GUI provider for Windows and/or WPF, a Selenium provider, etc.

Second point is that we don't necessarily __only__ want a screen shot. This is more of a "Collect failure details" call than just take a screenshot. Interesting. That implies that multiple providers may be installed to contribute. It does sound a lot like an event. 😄 

"The reasonable solution to my problem seems to be to only take a screenshot at the point of actual failure as that would only impact each failing test by only 1 second or so (and not every single test by hundreds of seconds). Selenium does not know when a test has "failed" so it seems reasonable this mechanism, on failure, would be provided by NUnit in some way." 

The point at which we know we have an actual failure is when the test exits. However, I suspect that isn't quite good enough, since your TearDown will most likely have disposed of resources at that point. In principle, the test result is supposed to be decided upon entry to teardown except, of course, that an unhandled exception can mess that up.

For that reason, it seems as if the event you need has to be triggered at the point where NUnit is about to run teardown. Alternatively, your teardown could detect that the test failed and actively invoke some nunit facility. However, that solution is not suitable for use in a base class, since teardowns of the derived class may have changed the status already.

NUnit V2 had a series of events in the life cycle of a test, which are currently mimicked by the nunit engine. But the engine is to far away to capture the sort of information you want.

@jnm2 Do you think we should revive the old listener addin for NUnit 3? With the addition of one for BareTestFinished (i.e. prior to teardown)?

**Comment by jnm2** [comment]
@CharliePoole I'd be interested in capturing multiple screenshots in tests that have multiple asserts, too. Can we make this possible?

**Comment by CharliePoole** [comment]
@jnm2 IMO we could just barely do this for the whole test using adhoc code. To do it for every assert result (perhaps even successful ones for some people) I would start by re-inventing the old NUnit addins for NUnit 3 - at least the test listener flavor. It just feels like you're crossing into the region where we need a bit more infrastructure.

**Comment by CharliePoole** [comment]
Thinking further... all the other uses of NUnit Addins have now been taken over by interfaces like `IFixtureBuilder`, `IApplyToTest`, etc. So what if we defined a listener interface? The key question is what would be put it on. An attribute doesn't seem to make a lot of sense here... or does it?

**Comment by jeremysimmons** [comment]
@jnm2 you might be interested in the Verification class https://github.com/ObjectivityLtd/Test.Automation/blob/master/Objectivity.Test.Automation.Common/Verify.cs to support a concept of "Multiple Asserts", including capturing multiple screenshots...

**Comment by uchagani** [comment]
Another vote on this.

My use case is more around unhandled exceptions that (it seems) nunit is catching. I don’t care much about assertion exceptions themselves but exceptions thrown by other code.
